category_encoders/utils.py

```diff
@@ -363,7 +363,7 @@ def get_feature_names(self) -> List[str]:
                       category=FutureWarning)
         return self.get_feature_names_out()
 
-    def get_feature_names_out(self) -> List[str]:
+    def get_feature_names_out(self, input_features=None) -> np.ndarray:
```

Comment:
so we're not actually doing anything with the `input_features` parameter? I thought the point of including it would be to explicitly tell every encoder how to calculate output features from the given input features (this is what the 1-1 mixin does, right?)  
For some encoder it give a not-fitted error, but other don't need fitting.  
I feel this is taking some shortcut that sklearn does not take

Reply:
Well, let me explain (for simplicity, every time I say a `Pipeline` it could be that or a `ColumnTransformer` or a `FeatureUnion`...): 
* We need to have the parameter `input_features` because otherwise, when the Encoder is part of a `Pipeline`, there will be an `Exception: BaseEncoder.get_feature_names_out() takes 1 positional argument but 2 were given`, because the method will be called with the parameter `input_features`. 
* Now, what can we do with this parameter? There were two options: modifying all encoders to take into account how to generate the output features given the input_features, or just using the `feature_names_out_`, which is available and computed. Given that this library internally work with dataframes, I thought that the second option was a much better idea. There is only one caveat, which happens again when the encoder is part of a `Pipeline`: if the previous transformer outputted a numpy array, then the column names won't be good (they will be like col_0_A, col_0_B...). That's why I think we should inform that, to ensure full compatibility and 100% correctness of feature names, `sklearn` should be used with the `set_output="pandas"` option. In this case, everything works smoothly, as the tests prove. 
* Another way to put it, in `sklearn` it makes sense to have a function with this signature, because an estimator can be fitted with a numpy array and then it should be able to "compute" the feature names of the output given a list of input_features. But for `category_encoders` this use case does not have too much sense, as the input is expected to be a pandas dataframe. And this assumption is spread everywhere: even other parameters work with column names, like the `mapping` or the `cols`; you won't see parameters expecting column names in `sklearn` because they expect that the input can be a numpy array. 
* So, yes, I agree the behavior is different than in `sklearn`, but I think it makes sense as long as the library is based on this "pandas assumption". I think a big redesign will be a good idea but as we discussed in the issue it will take a lot of effort.

Reply:
I agree that the proposed change is better than what we have at the moment since it fixes a bug.  
I've got one more question: If we change all encoders to properly work with the input features, will we get compatibility even without setting `set_output=pandas`? Even if we still do all the internal stuff in pandas? If so I think it would be worth it because we'd get compatibility even for sklearn versions < 1.2. This is in fact what I intended when I first raised the issue. In one of the comment I outlined what the changes for each encoder / family of encoders would be. This is far from a complete redesign of the library and also won't get rid of all dataframes. Does this make sense?  
I somehow really don't like introducing dependency versions in the code since this will make it a maintenance hell in the long run

Reply:
Well, my gut feeling is that it's not worth it: it will still be a very big effort and it will only fix one subtle bug (which doesn't make anything crash) and there will still be some inconsistencies. Let me try to explain... 

* **Current Situation**: If the `category_encoder` is inside a composed `sklearn` transfomer, which delegates the `get_feature_names_out` to every inner estimator, there is an Exception when calling `get_feature_names_out`, because the signature is not valid. The error is `TypeError: get_feature_names_out() takes 1 positional argument but 2 were given`. 
* **Super simple Fix**: By changing the signature but not doing anything with the input parameter, we remove this `Exception` and, making using of the attribute `feature_names_out_`, we get a right answer in most situations. I mean, will be right whenever the `category_encoder` is fitted and transformed alone, or whenever is included in a composed transformer but in the first layer. However, if the `category_encoder` is not in the first layer of a composed transformer, and if `set_output` is the default, then the answer will be wrong, because when fitted the previous transformer of the row will output a numpy array, and the `category_encoder` will take ["0", "1", ...] as the original column names. This is the only *issue* that will remain with this super simple fix (and the solution is as easy as setting `set_output=pandas`). 
* **More thorough fix**:  If we modify every encoder by adding a treatment of the input_features in order to generate the column names (which requieres some work for some of the encoders), the only added value will be removing the wrong answer in the mentioned case (when the encoder is part of a composed transformer and has another `sklearn` transfomer doing something before). My point is that, even in this situation, there will be some inconsistences that are related to the fact of assuming that the input is a pandas dataframe. One of those is the fact that some parameters of the encoders depend on column names. For example, you can specify the mapping for an `OrdinalEncoder` using this: 
    ```python
    OrdinalEncoder(mapping=[{'col': 'col_a', 'mapping': {'a': 2, 'b': 1, 'c': 3}}])
    ``` 
    This code will crash in the same situation as before: if the `OrdinalEncoder` is part o a pipeline and there is another `sklearn` transformer before, the columns received during `fit` won't correspond to the columns specified in the `mapping` parameter (unless `set_output=pandas`). You can take a look at this issue [here](https://colab.research.google.com/drive/1lJJvXyAOoG9MBtNiH2qSTrqbzZtEPTeB#scrollTo=nCJtM_1keNTD&line=2&uniqifier=1). That's why I was saying that if we wish to be fully aligned with `sklearn`, there will be more things to do (in particular, for example, no column names should be used in any parameter of any encoder). So I feel more inclined to say "all or nothing": if we want full compatibility with `sklearn`, that's fine, but it will be a breaking change a lots of work, and otherwise, I think the *super simple fix* is good enough. Well, that's my opinion...

Reply:
Sorry that I did not answer you question properly: regarding the versions, I would not specify the dependency of 1.2 in the requirements or the setup, because with the simple fix there the remaining bugs are not, in my opinion, a big issue. 

If you want, I could refactor the tests so that I don't require the version for running all tests, but only for some of the asserts (the one that will fail because the column names won't match what `get_feature_names_out` outputs). This way it will be more clear what does not work if not setting the output as pandas.

Reply:
> I've got one more question: If we change all encoders to properly work with the input features, will we get compatibility even without setting `set_output=pandas`?

Ok I see, this is not true since the columns to encode are always referenced by name in the `cols` parameter.  
Then I agree with you that changing `get_feature_names_out` to actually do something with the `input_features` is pretty much useless since it won't fix the issue.

Reply:
I pretty much agree with you that we should go for the super simple fix. Let me think a little about the version stuff, maybe there is another solution, but I think we're very close to getting this merged

Reply:
Hi, I've added some thoughts on the version. I'll resolve the discussion here. Please have a look at the other comment I've made and if you're fine with it we're ready to merge!

Reply:
Hi! Ok, I have added both suggestions!